File: golang/Set/set.go
package main

import (
	"fmt"
	"strings"
)

// Set 集合接口
type Set[E any] interface {
	Add(e E)
	Contains(e E) bool
	Remove(e E)
	GetSize() int
	IsEmpty() bool
}

// LinkListSet 基于链表的集合实现
type LinkListSet[E any] struct {
	list *LinkList[E]
}

// NewLinkListSet 创建一个新的链表集合
func NewLinkListSet[E any]() *LinkListSet[E] {
	return &LinkListSet[E]{
		list: NewLinkList[E](),
	}
}

func (s *LinkListSet[E]) GetSize() int {
	return s.list.GetSize()
}

func (s *LinkListSet[E]) IsEmpty() bool {
	return s.list.IsEmpty()
}

func (s *LinkListSet[E]) Add(e E) {
	// 只有当元素不存在时才添加
	if !s.list.Find(e) {
		s.list.AddFirst(e)
	}
}

func (s *LinkListSet[E]) Contains(e E) bool {
	return s.list.Find(e)
}

func (s *LinkListSet[E]) Remove(e E) {
	s.list.Remove(e)
}

func (s *LinkListSet[E]) String() string {
	var builder strings.Builder
	builder.WriteString("LinkListSet{")
	builder.WriteString(s.list.String())
	builder.WriteString("}")
	return builder.String()
}

func main() {
	fmt.Println("=== Go语言集合示例 ===")

	set := NewLinkListSet[int]()

	// 添加元素
	fmt.Println("--- 添加元素 ---")
	set.Add(1)
	set.Add(2)
	set.Add(3)
	set.Add(1) // 重复元素，不会被添加
	fmt.Printf("集合: %s\n", set.String())
	fmt.Printf("大小: %d\n", set.GetSize())

	// 查找元素
	fmt.Println("\n--- 查找元素 ---")
	fmt.Printf("包含2: %t\n", set.Contains(2))
	fmt.Printf("包含5: %t\n", set.Contains(5))

	// 删除元素
	fmt.Println("\n--- 删除元素 ---")
	set.Remove(2)
	fmt.Printf("删除2后: %s\n", set.String())
	fmt.Printf("大小: %d\n", set.GetSize())

	// 添加字符串
	fmt.Println("\n--- 字符串集合 ---")
	strSet := NewLinkListSet[string]()
	strSet.Add("Hello")
	strSet.Add("World")
	strSet.Add("Go")
	fmt.Printf("字符串集合: %s\n", strSet.String())

	fmt.Println("\n=== 示例程序结束 ===")
}

File: python/set.py
"""
Set Implementation in Python
Based on Java Set implementation
"""

from typing import TypeVar, Generic, Protocol, runtime_checkable
from abc import ABC, abstractmethod
from linklist import LinkList

T = TypeVar('T')


class Set(Generic[T], ABC):
    """Set interface"""

    @abstractmethod
    def add(self, e: T) -> None:
        """Add element to set"""
        pass

    @abstractmethod
    def contains(self, e: T) -> bool:
        """Check if element exists in set"""
        pass

    @abstractmethod
    def remove(self, e: T) -> None:
        """Remove element from set"""
        pass

    @abstractmethod
    def get_size(self) -> int:
        """Get set size"""
        pass

    @abstractmethod
    def is_empty(self) -> bool:
        """Check if set is empty"""
        pass


class LinkListSet(Set[T]):
    """Set implementation using linked list"""

    def __init__(self):
        self._list: LinkList[T] = LinkList[T]()

    def get_size(self) -> int:
        return self._list.get_size()

    def is_empty(self) -> bool:
        return self._list.is_empty()

    def add(self, e: T) -> None:
        """Add element to set (no duplicates)"""
        if not self._list.find(e):
            self._list.add_first(e)

    def contains(self, e: T) -> bool:
        return self._list.find(e)

    def remove(self, e: T) -> None:
        self._list.remove_element(e)

    def __str__(self) -> str:
        return f"LinkListSet({{{self._list.__str__()}}})"

    def __repr__(self) -> str:
        return self.__str__()

    def __len__(self) -> int:
        return self.get_size()

    def __contains__(self, item: T) -> bool:
        return self.contains(item)

    def __iter__(self):
        return iter(self._list)


if __name__ == "__main__":
    # Test LinkListSet
    print("Testing LinkListSet:")
    s = LinkListSet[int]()
    s.add(1)
    s.add(2)
    s.add(3)
    s.add(1)  # Duplicate, won't be added
    print(f"Set: {s}")
    print(f"Size: {s.get_size()}")
    print(f"Contains 2: {s.contains(2)}")
    print(f"Contains 5: {s.contains(5)}")

    s.remove(2)
    print(f"After remove 2: {s}")

    # Test iteration
    print("Iteration:")
    for item in s:
        print(item, end=" ")
    print()
